Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/ThresholdedRandomCutForest.java

```diff
@@ -432,6 +442,134 @@ public void setInitialThreshold(double initial) {
+            // first populate the description with current knowledge
+            // then update the preprocessor
+            // then update the RCF
+            if (result.getAnomalyGrade() > 0) {
+                /**
+                 * adds information of expected point to the result descriptor (provided it is
+                 * marked anomalous) Note that is uses relativeIndex; that is, it can determine
+                 * that the anomaly occurred in the past (but within the shingle) and not at the
+                 * current point -- even though the detection has triggered now While this may
+                 * appear to be improper, information theoretically we may have a situation
+                 * where an anomaly is only discoverable after the "horse has bolted" -- suppose
+                 * that we see a random mixture of the triples { 1, 2, 3} and {2, 4, 5}
+                 * corresponding to "slow weeks" and "busy weeks". For example 1, 2, 3, 1, 2, 3,
+                 * 2, 4, 5, 1, 2, 3, 2, 4, 5, ... etc. If we see { 2, 2, X } (at positions 0 and
+                 * 1 (mod 3)) and are yet to see X, then we can infer that the pattern is
+                 * anomalous -- but we cannot determine which of the 2's are to blame. If it
+                 * were the first 2, then the detection is late. If X = 3 then we know it is the
+                 * first 2 in that unfinished triple; and if X = 5 then it is the second 2. In a
+                 * sense we are only truly wiser once the bolted horse has returned! But if we
+                 * were to say that the anomaly was always at the second 2 then that appears to
+                 * be suboptimal -- one natural path can be based on the ratio of the triples {
+                 * 1, 2, 3} and {2, 4, 5} seen before. Even better, we can attempt to estimate a
+                 * dynamic time dependent ratio -- and that is what RCF would do.
+                 *
+                 * @param result the description of the current point
+                 */
+                int shingleSize = result.getShingleSize();
+                int dimension = result.getDimension();
+                int base = dimension / shingleSize;
+                double[] reference = result.getCurrentInput();
+                float[] newPoint = result.getExpectedRCFPoint();
+
+                int index = result.getRelativeIndex();
+                if (index < 0) {
+                    reference = preprocessor.getShingledInput(shingleSize + index);
+                    result.setPastTimeStamp(preprocessor.getTimeStamp(shingleSize + index));
+                }
+                result.setPastValues(reference);
+                if (newPoint != null) {
+                    double[] values = preprocessor.getExpectedValue(index, reference, point, newPoint);
+                    if (forestMode == ForestMode.TIME_AUGMENTED) {
+                        int endPosition = (shingleSize - 1 + index + 1) * dimension / shingleSize;
```

Comment:
Yes - but I kept the +1 and -1 so that it fits the pattern with the non time-augmented case. There were a bunch of errors while changing the code -0- this is redundant but maybe more explanatory.

Reply:
Ok - changing -- the entire block has a lot of "explanatory" repetition ....